orchagent/bfdorch.cpp

```diff
@@ -152,8 +155,52 @@ void BfdOrch::doTask(NotificationConsumer &consumer)
     }
 }
 
+bool BfdOrch::register_bfd_state_change_notification(void)
+{
+    sai_attribute_t  attr;
+    sai_status_t status;
+    sai_attr_capability_t capability;
+
+    status = sai_query_attribute_capability(gSwitchId, SAI_OBJECT_TYPE_SWITCH, 
+                                            SAI_SWITCH_ATTR_BFD_SESSION_STATE_CHANGE_NOTIFY,
+                                            &capability);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Unable to query the BFD change notification capability");
+        return false;
+    }
+
+    if (!capability.set_implemented)
+    {
+        SWSS_LOG_ERROR("BFD register change notification not supported");
+        return false;
+    }
+
+    attr.id = SAI_SWITCH_ATTR_BFD_SESSION_STATE_CHANGE_NOTIFY;
+    attr.value.ptr = (void *)on_bfd_session_state_change;
+
+    status = sai_switch_api->set_switch_attribute(gSwitchId, &attr);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to register BFD notification handler");
+        return false;
+    }
+    return true;
```

Comment:
should we have here? 'register_state_change_notif = true' why only on create session and not just when registration is done?

Reply:
Registration is done as part of first session creation just before we create BFD. 
When the user tries to create BFD session, registration support is checked and if not implemented a syslog error would be thrown.
If it is supported registration would be done and the first session would be created. Subsequent session creations will not perform registrations.